Put lock contention logging behind `DEBUG_LOCKCONTENTION` preprocessor directive

jonatack opened the pull request:
PRs #22736, #22904 and #23223 changed lock contention logging from a `DEBUG_LOCKCONTENTION` compile-time preprocessor directive to a runtime `lock` log category and improved the logging output. This changed the locking from using `lock()` to `try_lock()`:

- `void Mutex::UniqueLock::lock()` acquires the mutex and blocks until it gains access to it

- `bool Mutex::UniqueLock::try_lock()` doesn't block but instead immediately returns whether it acquired the mutex; it may be used by `lock()` internally as part of the deadlock-avoidance algorithm

In theory the cost of `try_lock` might be essentially the [same](https://www.erisian.com.au/bitcoin-core-dev/log-2022-03-31.html#l-697) relative to `lock`. The test-and-set logic of these calls is purported to be ~ constant time, optimised and light/quick if used carefully (i.e. no mutex convoying), compared to system calls, memory/cache coherency and fences, wait queues, and (particularly) lock contentions. See the discussion around https://github.com/bitcoin/bitcoin/pull/22736#issuecomment-902851054 and after with respect to performance/cost aspects.  However, there are reasonable concerns (see [here](https://github.com/bitcoin/bitcoin/pull/22736#discussion_r691277896) and [here](https://www.erisian.com.au/bitcoin-core-dev/log-2022-03-31.html#l-620)) that `Base::try_lock()` may be potentially [costly](https://www.erisian.com.au/bitcoin-core-dev/log-2022-03-31.html#l-700) or [risky](https://github.com/bitcoin/bitcoin/pull/22904#issuecomment-930484001) compared to `Base::lock()` in this very frequently called code.

One alternative to keep the run-time lock logging would be to gate the `try_lock` call behind the logging conditional, for example as proposed in https://github.com/bitcoin/bitcoin/commit/ccd73de1dd969097d34634c2be2fc32b03fbd09e and ACKed [here](https://github.com/bitcoin/bitcoin/pull/22736#issuecomment-901980815). However, this would add the [cost](https://github.com/bitcoin/bitcoin/pull/22736#issuecomment-910102353) of `if (LogAcceptCategory(BCLog::LOCK))` to the hotspot, instead of replacing `lock` with `try_lock`, for the most frequent happy path (non-contention).

It turns out we can keep the advantages of the runtime lock contention logging (the ability to turn it on/off at runtime) while out of prudence putting the `try_lock()` call and `lock` logging category behind a  `DEBUG_LOCKCONTENTION` compile-time preprocessor directive, and also still retain the lock logging enhancements of the mentioned PRs, as suggested in https://github.com/bitcoin/bitcoin/pull/24734#issuecomment-1085785480 by W. J. van der Laan, in https://github.com/bitcoin/bitcoin/pull/22736#discussion_r691280693, and in the linked IRC discussion.

Proposed here and for backport to v23.

jonatack committed:
Put lock logging behind DEBUG_LOCKCONTENTION preprocessor directive

jonatack commented on src/sync.h:
```diff
@@ -136,9 +136,14 @@ class SCOPED_LOCKABLE UniqueLock : public Base
-        Base::lock();
+#ifdef DEBUG_LOCKCONTENTION
+        if (!Base::try_lock()) {
+            LOG_TIME_MICROS_WITH_CATEGORY(strprintf("lock contention %s, %s:%d", pszName, pszFile, nLine), BCLog::LOCK);
```
```suggestion
            LOG_TIME_MICROS_WITH_CATEGORY(strprintf("lock contention %s, %s:%d", pszName, pszFile, nLine));
```

ajtowns replied:
```c++
#ifdef DEBUG_LOCKCONTENTION
        if (Base::try_lock()) return;
        LOG_TIME_MICROS(....);
#endif
        Base::lock();
````

jonatack replied:
Agree, more like the current code.

jonatack replied:
Done, thanks

jonatack commented:
> Alternatively, we could keep the lock category only when `DEBUG_LOCKCONTENTION` is set. This still has the intended effect of making lock contention run-time switchable for people that want it enabled. No strong opinion though. 

Agree! This would allow leaving the flag defined for development while being able to switch it off at runtime. All the benefits without the drawbacks. Will update.

jonatack commented:
Updated code and PR title/description per @laanwj's suggestion. This is now a much smaller change, too.

jonatack commented:
Updated code and PR title/description per @laanwj's and @ajtowns suggestion. This is now a much smaller change, too.

shaavan reviewed (approved):
ACK 73759f88537e472dc443318a16545d63e6c22024

Let me summarize how I understood this PR:

- This PR introduced (back) the `DEBUG_LOCKCONTENTION` compile-time processor directive to allow the user to opt-out of potentially costly and dangerous run time locking log using `Base::try_lock()`.

If my understanding of this PR is correct, I agree with its concept of it:

- I like the prudent approach in case we find out using `Base::try_lock()` was not the efficient way of logging.
- This keeps the earlier benefits introduced with run-time logging, so we are not losing a beneficial feature.

Considering the comments of @laanwj and @ajtowns

- The reduced diff looks much cleaner than the earlier version of this PR.
- As per @laanwj suggestion, the updated code keeps the benefits of run-time logging instead of completely reverting them, which I agree with.

jonatack commented:
Mind that updating the checkqueue_tests setup doesn't seem needed, but can do if people prefer.

MarcoFalke commented on test/functional/rpc_misc.py:
```diff
@@ -57,7 +57,7 @@ def run_test(self):
 
         # Test logging RPC returns the expected number of logging categories.
-        assert_equal(len(node.logging()), 27)
+        assert 26 <= len(node.logging()) <= 27
```
What is the point of this other than causing merge conflicts? Might be best to remove.

jonatack replied:
Considered removing, done

jonatack commented:
With this change, if `lock` logging is requested and "DEBUG_LOCKCONTENTION" is not defined:

$ bitcoind -debug=lock
Warning: Unsupported logging category -debug=lock.
2022-04-01T13:57:44Z [init] Bitcoin Core version v23.99.0-73759f88537e (debug build)
2022-04-01T13:57:44Z [init] Warning: Unsupported logging category -debug=lock.
```
```
$ bitcoin-cli logging '["lock"]'
error code: -8
error message:
unknown logging category lock
```

MarcoFalke commented:
So this will be disabled for all CI tasks? If yes, could enable it for some, or by default in `--enable-debug`? See also #24709.

Otherwise the code is never compiled, which makes me wonder if it ever was compiled before the changes?

MarcoFalke commented:
If it was never enabled previously, it can also be done as a follow-up. Just wanted to raise the point.

jonatack commented:
Follow-up SGTM. Right, the previous situation AFAIK was that it was only compiled if manually defined by the user.

jonatack commented:
Made the `src/sync.h` logging includes conditional on `DEBUG_LOCKCONTENTION` as well.

jonatack committed:
net: add missing include headers to src/net.{h,cpp}

Among these includes, src/net.{h,cpp} is relying on src/sync.h for <logging.h>
and in turn on src/logging.h for <list>.  The compiler raises when they are
removed from src/sync.h in the last commit of this PR, so they need to be added.

jonatack committed:
refactor: move StartExtraBlockRelayPeers from net.h to net.cpp

which allows dropping `#include <logging>` in the header file.

jonatack committed:
Add DEBUG_LOCKCONTENTION documentation to the developer notes

jonatack commented:
Added developer notes documentation.

laanwj commented on test/functional/rpc_misc.py:
```diff
@@ -56,9 +56,6 @@ def run_test(self):
 
         self.log.info("test logging rpc and help")
 
-        # Test logging RPC returns the expected number of logging categories.
```
ACK on removing this test, it seems kind of unnecessary (just checks that the code is the code, not any kind of correctness).

jonatack replied:
Agree

laanwj commented:
Concept and code review ACK 2623fb4ff235b199e167695fe304a689b072c560

jonatack commented:
> (reasoning for adding backport label is that it removes the `try_lock` → `lock` sequence overhead for every locking in the application, which adds some overhead and is potentially risky)

Agree, really happy that this change avoids unnecessary risk while preserving the benefits of the recent improvements.

jonatack commented:
(This could maybe use a better tag than `Tests`, if there is a more applicable one.)

jonatack commented:
> So this will be disabled for all CI tasks? If yes, could enable it for some, or by default in `--enable-debug`? See also #24709.

Done in #24757.

fanquake commented on src/net.cpp:
```diff
@@ -51,8 +52,12 @@
 #include <array>
 #include <cstdint>
 #include <functional>
+#include <list>
```
In 6e56589bcbe2d04ee1b9018baf5a9d1dc958e68d

> The compiler raises when they are removed from src/sync.h in the last commit of this PR, so they need to be added.

If they need to be added, they should be added in the commit where they become required. If this PR is going to be backported, the number of commits, and potential for conflicts should be kept at a minimum, so we should avoid things like additional commits doing include fixups.

jonatack replied:
These are different changes; kept them separate to ease review. If the goal is to squash this down, then we could not conditionally do the two includes in the 3rd commit and backport only that one.

jonatack replied:
Minimal version in #24770 for backport.

fanquake commented on src/net.cpp:
```diff
@@ -1862,6 +1862,12 @@ void CConnman::SetTryNewOutboundPeer(bool flag)
     LogPrint(BCLog::NET, "net: setting try another outbound peer=%s\n", flag ? "true" : "false");
 }
 
+void CConnman::StartExtraBlockRelayPeers()
```
In e1437e1f5d3fb9ae95b1ca503893ff0ffb2e8c8d
> which allows dropping `#include <logging>` in the header file.

is this the only reason for this change? I'm not sure that doing this is required right now, and it seems to just be what necessitates the changes in the prior commit.

jonatack replied:
No; see explanation in the first commit message.

jonatack commented:
> ACK [73759f8](https://github.com/bitcoin/bitcoin/commit/73759f88537e472dc443318a16545d63e6c22024)

@shaavan mind having another look? (thanks!)

fanquake reviewed (commented):
I think you could just drop the first two commits, and add 2 missing includes to the 3rd commit, i.e https://github.com/fanquake/bitcoin/tree/simplfied_24734. That would seem to achieve the same outcome with less code changes, and is simpler if this is meant to be backported.

shaavan reviewed (commented):
Changes since my last review:

1. Added documentation for `DEBUG_LOCKCONTENTION` in developer notes.
2. Added missing include headers.
3. Moved `StartExtraBlockRelayPeers` from net.h to the net.cpp

I like the added developer documentation for its clear and concise writing. It explains the purpose of the `DEBUG_LOCKCONTENTION` macro and describes how to use it effectively.

However, I am not able to reason with the following added changes:

- **Keeping all the included headers into a separate commit.**
    - Though it makes reviewing easier. I don’t feel it needs to be in a separate commit.
- **Moving `StartExtraBlockRelayPeers` from net.h to the net.cpp**
    - Though it is mentioned in [this](https://github.com/bitcoin/bitcoin/pull/24734#discussion_r842591735) comment to refer to the first commit message, I can still not completely understand the reasoning. Hence, for now, I am concept 0 on this particular change.

I took a look into #24770, which has all the changes I agree with from this PR, and hence I think we can focus on getting that merged first.

jonatack commented:
@shaavan thanks for having a look.

>     * **Keeping all the included headers into a separate commit.**
>       
>       * Though it makes reviewing easier. I don’t feel it needs to be in a separate commit.

The missing headers are an unrelated issue, apart from needing to be fixed for the object of this pull. Missing headers are generally done in a separate commit, see https://github.com/bitcoin/bitcoin/pull/24740/commits for a recent example.
 
>     * **Moving `StartExtraBlockRelayPeers` from net.h to the net.cpp**
>       
>       * Though it is mentioned in [this](https://github.com/bitcoin/bitcoin/pull/24734#discussion_r842591735) comment to refer to the first commit message, I can still not completely understand the reasoning. Hence, for now, I am concept 0 on this particular change.

Moving the function definition into the implementation (cpp) allows dropping an include header. Doing so can speed up compilation for the many files that include net.h. It's not strictly needed for backport but is trivial to review.

ajtowns commented:
This should be closed or rebased on top of #24770 if that's what you're now intending to get merged...

jonatack commented:
Yes, this is my preferred version and plan to update or close whichever ones aren't merged.

shaavan reviewed (approved):
ACK 7ea951d607846f21c262aca950c393cc150ca7ab

> Missing headers are generally done in a separate commit.
    
Though I am not 100% convinced, I cannot think of any blocking points for this change. Hence I am Concept 0 on this and will allow fellow reviewers to decide the optimal way forward on this particular change.
    
> Moving the function definition into the implementation (cpp) allows dropping an include header. Doing so can speed up compilation for the many files that include the net.h.
    
That is acceptable reasoning, and I find it to be logically sound. Hence, I agree with this change.

MarcoFalke commented:
Closing as this has been merged

jonatack commented:
Thank you for revewing, @shaavan. I moved the remaining changes here to #24757.
